src/Build/BackEnd/Components/RequestBuilder/Lookup.cs

```diff
@@ -658,17 +659,41 @@ internal void AddNewItemsOfItemType(string itemType, ICollection<ProjectItemInst
             IEnumerable<ProjectItemInstance> itemsToAdd = group;
             if (doNotAddDuplicates)
             {
-                // Remove duplicates from the inputs.
-                itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
 
                 // Ensure we don't also add any that already exist.
                 var existingItems = GetItems(itemType);
+              
+                var existingItemsHashSet = existingItems.ToHashSet(ProjectItemInstance.EqualityComparer);
+
                 if (existingItems.Count > 0)
                 {
-                    itemsToAdd = itemsToAdd.Where(item => !existingItems.Contains(item, ProjectItemInstance.EqualityComparer));
+
+                    var deduplicatedItemsToAdd = new List<ProjectItemInstance>();
+                    foreach (var item in itemsToAdd)
+                    {
+                        if (existingItemsHashSet.Add(item))
+                        {
+                            deduplicatedItemsToAdd.Add(item);
+                        }
+                    }
+                    itemsToAdd = deduplicatedItemsToAdd;
+                }
+                else
+                {
+                    // Remove the duplicates in case we're not concerned with the existing items.
+                    itemsToAdd = itemsToAdd.Distinct(ProjectItemInstance.EqualityComparer);
                 }
             }
 
+            if (doNotAddDuplicates)
+            {
+                logFunction?.Invoke(itemsToAdd.ToList());
+            }
+            else {
+                var groupAsList = group as List<ProjectItemInstance>;
```

Comment:
will this group ever be a List<T>? or never? or sometimes?

Reply:
I double-checked, and this function is only called in two places, and in both of them it gets passed a `List<T>`:
https://source.dot.net/#Microsoft.Build/BackEnd/Components/RequestBuilder/Lookup.cs,35477908c518ae2a,references

I also debugged and verified that it's List<T> in both cases (ExpandIntoItemsLeaveEscaped always allocates and returns a `List<T>`). 

So in theory we could remove the cast and type check, but I'd still keep it just in case things change in the future. But for our purposes today it's helpful to know that the `group` is always a `List<T>`.

Reply:
yes. I meant it there as a future proofing since the function does not guarantee that it will only ever be a list so it could introduce an annoying bug down the line.